File: app/src/main/java/me/blog/korn123/easydiary/adapters/FontItemAdapter.java
package me.blog.korn123.easydiary.adapters;

import android.app.Activity;
import android.content.Context;
import android.graphics.BitmapFactory;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import org.apache.commons.lang3.StringUtils;

import java.util.List;
import java.util.Map;

import me.blog.korn123.commons.constants.Constants;
import me.blog.korn123.commons.utils.CommonUtils;
import me.blog.korn123.commons.utils.FontUtils;
import me.blog.korn123.easydiary.R;

/**
 * Created by CHO HANJOONG on 2017-03-16.
 */

public class FontItemAdapter extends ArrayAdapter<Map<String, String>> {
    private final Context context;
    private final List<Map<String, String>> list;
    private final int layoutResourceId;

    public FontItemAdapter(Context context, int layoutResourceId, List<Map<String, String>> list) {
        super(context, layoutResourceId, list);
        this.context = context;
        this.list = list;
        this.layoutResourceId = layoutResourceId;
    }

    public View getView(int position, View convertView, ViewGroup parent) {
        View row = convertView;
        ViewHolder holder = null;
        if (row == null) {
            LayoutInflater inflater = ((Activity)this.context).getLayoutInflater();
            row = inflater.inflate(this.layoutResourceId, parent, false);
            holder = new ViewHolder();
            holder.textView = (TextView) row.findViewById(R.id.textView);
            holder.imageView = (ImageView) row.findViewById(R.id.checkIcon);
            row.setTag(holder);
        } else {
            holder = (ViewHolder)row.getTag();
        }

        if (StringUtils.equals(CommonUtils.loadStringPreference(context, Constants.SETTING_FONT_NAME, Constants.CUSTOM_FONTS_SUPPORTED_LANGUAGE_DEFAULT), list.get(position).get("fontName"))) {
            holder.imageView.setImageBitmap(BitmapFactory.decodeResource(context.getResources(), R.drawable.check_mark));
        } else {
            holder.imageView.setImageBitmap(BitmapFactory.decodeResource(context.getResources(), R.drawable.check_mark_off));
        }
        holder.textView.setText(list.get(position).get("disPlayFontName"));
        holder.textView.setTypeface(FontUtils.getTypeface(context, context.getAssets(), list.get(position).get("fontName")));

        return row;
    }

    private static class ViewHolder {
        TextView textView;
        ImageView imageView;
    }
}

File: app/src/main/java/me/blog/korn123/easydiary/adapters/FontItemAdapter.kt
package me.blog.korn123.easydiary.adapters

import android.app.Activity
import android.graphics.BitmapFactory
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import kotlinx.coroutines.*
import me.blog.korn123.commons.utils.EasyDiaryUtils
import me.blog.korn123.commons.utils.FontUtils
import me.blog.korn123.easydiary.R
import me.blog.korn123.easydiary.extensions.config
import me.blog.korn123.easydiary.extensions.initTextSize
import me.blog.korn123.easydiary.extensions.updateDrawableColorInnerCardView
import me.blog.korn123.easydiary.extensions.updateTextColors
import me.blog.korn123.easydiary.helper.AAF_TEST
import org.apache.commons.lang3.StringUtils

/**
 * Created by CHO HANJOONG on 2017-03-16.
 * Refactored code on 2019-12-26.
 *
 */

class FontItemAdapter(val activity: Activity, private val layoutResourceId: Int, private val list: List<Map<String, String>>
) : ArrayAdapter<Map<String, String>>(activity , layoutResourceId, list) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val itemView: View = convertView ?: LayoutInflater.from(parent.context).inflate(this.layoutResourceId, parent, false)
        if (itemView is ViewGroup) {
            activity.run {
                initTextSize(itemView)
                updateTextColors(itemView)
            }
        }

        when (itemView.tag is ViewHolder) {
            true -> itemView.tag as ViewHolder
            false -> {
                val viewHolder = ViewHolder(itemView.findViewById(R.id.textView), itemView.findViewById(R.id.checkIcon))
                itemView.tag = viewHolder
                viewHolder
            }
        }.run {
            if (StringUtils.equals(context.config.settingFontName, list[position]["fontName"])) {
                activity.updateDrawableColorInnerCardView(R.drawable.ic_check_mark)
                ContextCompat.getDrawable(context, R.drawable.ic_check_mark).run {
                    imageView.setImageDrawable(this)
                    imageView.alpha = 1F
                }
            } else {
                imageView.setImageBitmap(BitmapFactory.decodeResource(context.resources, R.drawable.ic_check_mark_off))
                imageView.alpha = 0.1F
            }
            textView.text = StringUtils.EMPTY
            Log.i(AAF_TEST, "${this.position} -> $position")
//        holder.textView?.typeface = FontUtils.getTypeface(context, context.assets, list[position]["fontName"])

            this.position = position
//        Log.i("fontDialog", "$position")
            renderJob(activity, this, position)
        }
        return itemView
    }

    class ViewHolder(val textView: TextView, val imageView: ImageView, var position: Int = 0)

    inner class FontItemRenderer(val activity: Activity, val holder: ViewHolder, val position: Int) : Thread() {
        override fun run() {
            when (holder.position == position) {
                true -> {
                    holder.textView.run {
                        val tf = FontUtils.getTypeface(context, list[position]["fontName"])
                        activity.runOnUiThread {
                            typeface = tf
                            text = list[position]["disPlayFontName"]
                        }
                    }
                    Log.i(AAF_TEST, "$position End")
                }
                false -> { Log.i(AAF_TEST, "$position Cancel") }
            }
        }
    }

    private fun renderJob(activity: Activity, holder: ViewHolder, position: Int) {
        Log.i(AAF_TEST, "$position Start")
//        FontItemRenderer(activity, holder, position).apply { start() }

        CoroutineScope(Dispatchers.IO).launch {
            when (holder.position == position) {
                true -> {
                    holder.textView.run {
                        val tf = FontUtils.getTypeface(context, list[position]["fontName"])
                        withContext(Dispatchers.Main) {
                            val label = if (activity.config.enableDebugMode) "\uD83C\uDF0E\uD83D\uDCF1${list[position]["disPlayFontName"]}" else list[position]["disPlayFontName"]
                            typeface = tf
                            text = label
                        }
                        if (activity.config.enableDebugMode) EasyDiaryUtils.highlightString(this)
                    }
                    Log.i(AAF_TEST, "${holder.position} End")
                }
                false -> { Log.i(AAF_TEST, "$position Cancel") }
            }
        }
    }
}
